signac/contrib/job.py

```diff
@@ -359,6 +359,14 @@ def _statepoint_filename(self):
         # use str.join with os.sep instead of os.path.join for speed.
         return os.sep.join((self.workspace(), self.FN_MANIFEST))
 
+    # TODO: Why is ws a property while workspace is a method? Why don't the
+    # Project methods for workspace/root_directory also behave the same? We
+    # should standardize this.
+    # TODO: The discrepancy between Project and Job in what is defined as a
+    # "workspace" is also slightly confusing. In view of moving towards a
+    # unified 3.0 API where both are subclasses of a common Directory or
+    # similar, we should aim to unify these APIs (although probably with an
+    # extremely long deprecation period like the entire lifetime of signac 2.0)
```

Comment:
Right, my understanding is that workspace being a function is a historical artifact rather than an indication of how we would want to design in now. That is addressing the first note here, but not the second. Apologies for sticking two separate ideas immediately next to each other like this.

The second question I'm asking here is, rather than converting workspace into a property, should we consider defining a new property `Job.root` (or `Job.path` or similar), mimicking it in `Project`, and then deprecating `Job.workspace`, `Job.ws`, and `Project.root_directory`? The reason I ask this is that we currently have different names for the path to a `Job` and the path to a `Project`, and even worse, `Project.workspace` means something semantically different from `Job.workspace`. The historical reasons for these names make sense, but now would be a good time to try and choose something more canonically appealing like `path`. `Project.workspace` would still exist since that has a specific meaning in the context of signac's (current) data model, but the other properties would then be aligned.

This change would also be in line with our original signac 2.0 proposal to enable different folder layouts via different data models, which would be greatly facilitated by moving towards a common `signac.Directory` class as a common parent for `Job` and `Project` that implements this `path`/`root` interface.

Reply:
I could buy into the `signac.Directory` (abstract?) class providing a `path` property.

Regardless of naming or plans for 3.0, we should definitely change the methods that should be properties in 2.0.

Reply:
I've outlined all of these ideas in the 2.0 API I proposed at some point. The unification of Project and Job to Directory, those being basically Paths with some additional properties and so on. Are those concepts considered here? Should I link them again?

Reply:
I’d appreciate a link. It’s been a long time since I’ve seen that document and I’m not sure where it lives.

Reply:
> Are those concepts considered here? Should I link them again?

@csadorf That's what I was referring to with this:

> This change would also be in line with our original signac 2.0 proposal to enable different folder layouts via different data models, which would be greatly facilitated by moving towards a common signac.Directory class as a common parent for Job and Project that implements this path/root interface.

Is that not what you're asking about? The whole point of my proposing these changes now is to make the necessary breaking changes in 2.0 so that we could implement (at least some of) the components of the original 2.0 proposal in minor releases along the 2.x line rather than needing to make a breaking major release.

@bdice here's [the link](https://github.com/glotzerlab/signac-2-prototype/). I intended to bring up long-term roadmap discussions at the next dev meeting anyway, so I can outline some ideas when I do.

Reply:
> Is that not what you're asking about? The whole point of my proposing these changes now is to make the necessary breaking changes in 2.0 so that we could implement (at least some of) the components of the original 2.0 proposal in minor releases along the 2.x line rather than needing to make a breaking major release.

@vyasr Yeah, I think I got my wires crossed on this one.